farm/modeling/biadaptive_model.py

```diff
@@ -0,0 +1,537 @@
+import copy
+import json
+import logging
+import os
+from argparse import Namespace
+from pathlib import Path
+
+import multiprocessing
+import numpy
+import torch
+from torch import nn
+
+from farm.data_handler.processor import TextSimilarityProcessor
+from farm.data_handler.data_silo import DataSilo
+from farm.modeling.language_model import LanguageModel
+from farm.modeling.prediction_head import PredictionHead, TextSimilarityHead
+from farm.modeling.tokenization import Tokenizer
+from farm.utils import MLFlowLogger as MlLogger, stack
+
+logger = logging.getLogger(__name__)
+
+
+class BaseBiAdaptiveModel:
+    """
+    Base Class for implementing AdaptiveModel with frameworks like PyTorch and ONNX.
+    """
+
+    subclasses = {}
+
+    def __init_subclass__(cls, **kwargs):
+        """ This automatically keeps track of all available subclasses.
+        Enables generic load() for all specific AdaptiveModel implementation.
+        """
+        super().__init_subclass__(**kwargs)
+        cls.subclasses[cls.__name__] = cls
+
+    def __init__(self, prediction_heads):
+        self.prediction_heads = prediction_heads
+
+    @classmethod
+    def load(cls, **kwargs):
+        """
+        Load corresponding AdaptiveModel Class(AdaptiveModel/ONNXAdaptiveModel) based on the
+        files in the load_dir.
+
+        :param kwargs: arguments to pass for loading the model.
+        :return: instance of a model
+        """
+        if (Path(kwargs["load_dir"]) / "model.onnx").is_file():
+            model = cls.subclasses["ONNXBiAdaptiveModel"].load(**kwargs)
+        else:
+            model = cls.subclasses["BiAdaptiveModel"].load(**kwargs)
+        return model
+
+    def logits_to_preds(self, logits, **kwargs):
+        """
+        Get predictions from all prediction heads.
+
+        :param logits: logits, can vary in shape and type, depending on task
+        :type logits: object
+        :param label_maps: Maps from label encoding to label string
+        :param label_maps: dict
+        :return: A list of all predictions from all prediction heads
+        """
+        all_preds = []
+        # collect preds from all heads
+        for head, logits_for_head in zip(self.prediction_heads, logits):
+            preds = head.logits_to_preds(logits=logits_for_head, **kwargs)
+            all_preds.append(preds)
+        return all_preds
+
+    def formatted_preds(self, logits, language_model1, language_model2, **kwargs):
+        """
+        Format predictions for inference.
+
+        :param logits: model logits
+        :type logits: torch.tensor
+        :param kwargs: placeholder for passing generic parameters
+        :type kwargs: object
+        :return: predictions in the right format
+        """
+        n_heads = len(self.prediction_heads)
+
+        if n_heads == 0:
```

Comment:
I think the cases for n_heads == 0 and n_heads == 2 are not working at the moment for a BiAdaptiveModel. We should then remove them here.